rllib/core/optim/examples/dummy_ppo_optimizer.py

```diff
@@ -0,0 +1,144 @@
+
+    def compute_loss(self, batch, fwd_out):
+        adv = batch["rewards"] - fwd_out["vf"]
+        actor_loss = -(fwd_out["logp"] * adv).mean()
+        critic_loss = (adv**2).mean()
+        loss = actor_loss + critic_loss
+        loss_dict = {
+            "total_loss": loss,
+            "actor_loss": actor_loss,
+            "critic_loss": critic_loss,
+        }
+        return loss_dict
+
+    def _configure_optimizers(self):
+        return [
+            torch.optim.SGD(self.module.parameters(), lr=self._config.get("lr", 1e-3))
+        ]
+
+    def get_state(self):
+        return [optim.state_dict() for optim in self.optimizers]
+
+    def set_state(self, state):
+        assert len(state) == len(self.optimizers)
+        for optim, optim_state in zip(self.optimizers, state):
+            optim.load_state_dict(optim_state)
+
+
+class DummyPPOTorchTrainer:
+    def __init__(self, config) -> None:
+        module_config = config["module_config"]
+        optimizer_config = {}
+        self._module = SimplePPOModule(module_config)
+        self._rl_optimizer = DummyPPOTorchOptimizer(self._module, optimizer_config)
+
+    @staticmethod
+    def on_after_compute_gradients(gradients_dict):
+        return gradients_dict
+
+    def compile_results(
+        self, batch, fwd_out, postprocessed_loss, post_processed_gradients
+    ):
+        """Compile results from the update."""
+
+        loss_numpy = convert_to_numpy(postprocessed_loss)
+        results = {
+            "avg_reward": batch["rewards"].mean(),
+        }
+        results.update(loss_numpy)
+        return results
+
+    def update(self, batch):
+        """Perform an update on this Trainer."""
+        torch_batch = convert_to_torch_tensor(batch)
+        fwd_out = self._module.forward_train(torch_batch)
+        loss = self._rl_optimizer.compute_loss(torch_batch, fwd_out)
+        postprocessed_loss = self._rl_optimizer.on_after_compute_loss(loss)
+        gradients = self.compute_gradients(loss)
+        post_processed_gradients = self.on_after_compute_gradients(gradients)
+        self.apply_gradients(post_processed_gradients)
+        results = self.compile_results(
+            batch, fwd_out, postprocessed_loss, post_processed_gradients
+        )
+        return results
+
+    def compute_gradients(self, optimization_vars, **kwargs):
+        """Perform an update on self._module
+
+            For example compute and apply gradients to self._module if
+                necessary.
+
+        Args:
+            optimization_vars: named variables used for optimizing self._module
+
+        Returns:
+            A dictionary of extra information and statistics.
+        """
+        # for torch
+        optimization_vars["total_loss"].backward()
+        grads = {n: p.grad for n, p in self._module.named_parameters()}
+        return grads
+
+    def apply_gradients(self, gradients):
+        """Perform an update on self._module"""
+        del gradients
```

Comment:
assert either grandients or self.optimizer exists?

Reply:
Sure? It's not relevant for this example script. I'm just doing this since the parameter is unused for torch but will be used for tensorflow.

Reply:
like in tensorflow you literally apply the gradients to a network. In torch theyre accumulated on gradient fields of tensors.